### PR TITLE
http/server: close connection on unhandled internal errors

### PR DESCRIPTION
**Pull request by pspacek**
Fixes: #146

I'm not sure if implementation is correct, or if it best way to handle
it. In any case it fixes client-visible problem where lua-http server
keep connection open even after internal error (like #145).

### CODE REVIEW

**Comment by pspacek** [comment]
Also, I have no idea how to write for this because it would require a way to cause exception from client side ... which is certainly something I do not want to see in server code ;-)

File: http/server.lua [hunk 1]
```diff
@@ -1,3 +1,4 @@
+local debug = require "debug"
```
Comment (daurnimator) [inline http/server.lua]: no need for this; debug is a standard lua library.

File: http/server.lua [hunk 2]
```diff
@@ -152,7 +153,16 @@ local function handle_socket(self, socket)
 		while true do
 			local timeout = deadline and deadline-monotime() or self.intra_stream_timeout
 			local stream
-			stream, err, errno = conn:get_next_incoming_stream(timeout)
+			local stream_or_err
+			local finished
+			finished, stream_or_err, err, errno = xpcall(conn.get_next_incoming_stream, function(err) return debug.traceback(err) end, conn, timeout)
```
Comment (daurnimator) [inline http/server.lua]: try and avoid `xpcall` on hot paths; it has quite an impact.

Also you can pass `debug.traceback` directly.

Comment (pspacek) [reply http/server.lua]: I'm not Lua expert - can you advise a different approach to catch errors instead of `xpcall`?

Comment (pspacek) [reply http/server.lua]: I wonder if it would be sufficient to wrap whole `while` cycle in `xpcall()`. Opinions?

**Comment by daurnimator** [comment]
See https://github.com/daurnimator/lua-http/issues/146#issuecomment-487392663

**Pull request closed without merging**